### PR TITLE
No sharing (collaborators) functionality for datasets

### PR DESCRIPTION
**Pull request by sroderick-g5sro**
Fixes: #247

### CODE REVIEW

**Comment by sroderick-g5sro** [comment]
@jonc125 I have tried to follow the instructions on the issue but have got stuck again - apologies. Please would you mind having a quick look at the following files to see if I am heading in the right direction:

dataset_header.html
datasets\views.py - DatasetCollaboratorsView
datasets\urls.py
templates\datasets\dataset_collaborators_form.html

File: weblab/datasets/views.py [hunk 1]
```diff
@@ -486,3 +487,10 @@ def post(self, request, *args, **kwargs):
             return HttpResponseRedirect(self.get_success_url())
         else:
             return self.render_to_response(self.get_context_data(forms=forms))
+
+
+class DatasetCollaboratorsView(EntityCollaboratorsView):
```
Comment (jonc125) [inline weblab/datasets/views.py:492]: Ah, I've realised an issue with just inheriting here. `EntityCollaboratorsView` inherits from `EntityTypeMixin` which we *don't* want to do. So possibly we need another mixin created in `entities/views.py` that contains the common fields/methods, and which both concrete views can inherit from.

So perhaps a new `EditCollaboratorsAbstractView` class, which contains everything from `EntityCollaboratorsView` except:
* no `template_name` field
* get_context_data is changed to remove the `kwargs['type'] = self.object.entity_type` line

Then we could have
```python
class EntityCollaboratorsView(EntityTypeMixin, EditCollaboratorsAbstractView):
    template_name = 'entities/entity_collaborators_form.html'

    def get_context_data(self, **kwargs):
        kwargs['type'] = self.object.entity_type
        return super().get_context_data(**kwargs)
```
and
```python
class DatasetCollaboratorsView(EditCollaboratorsAbstractView):
    template_name = 'datasets/dataset_collaborators_form.html'
```

File: weblab/templates/datasets/includes/dataset_header.html [hunk 2]
```diff
@@ -16,7 +16,8 @@ <h1 id="entityname">
         <img src="{% static 'img/delete.png' %}" alt="delete dataset"
              title="delete dataset"/></a>
     {% endif %}
-
+    <a title="manage collaborators for this dataset" href="{% url 'datasets:entity_collaborators' dataset.id %}">
+        Manage collaborators
```
Comment (jonc125) [inline weblab/templates/datasets/includes/dataset_header.html:20]: I think we need a `</a>` here, and possibly a full stop!

File: weblab/datasets/views.py (hunk 1 above)
Comment (sroderick-g5sro) [reply weblab/datasets/views.py:492]: thanks. Ive implemented that but I now get the following error:

`DatasetCollaboratorsView is missing a QuerySet. Define DatasetCollaboratorsView.model, DatasetCollaboratorsView.queryset, or override DatasetCollaboratorsView.get_queryset().`

Comment (jonc125) [reply weblab/datasets/views.py]: Ah yes, that's given by EntityTypeMixin for the other view; we need to set `model = Dataset` as a field for the datasets one.

**Comment by sroderick-g5sro** [comment]
I have now added in `get_context_data` and the form is working. However when I try and assign a dataset to another user I get a permission matching query does not exist error.

**Comment by jonc125** [comment]
Interesting, because using the add_collaborator method directly we’ve tested and it works. It’d definitely be worth creating a failing test that uses the new view to add a collaborator so we have something more straightforward to debug. Possibly going via the form (which lives in entities) has caused Django to get confused about which app it’s looking for the edit_entity permission in?

**Comment by sroderick-g5sro** [comment]
apologes @jonc125 I have attempted to write a unit test going via the form as below (so we can debug the permissions issue):

  ```
  def test_can_share_dataset(self, logged_in_user, other_user, public_protocol, client):
        shared_dataset = recipes.dataset.make(
            author=logged_in_user, name='mydataset', visibility='private', protocol=public_protocol)
        response = client.post('/datasets/%d/collaborators' % shared_dataset.pk,
                               {
                                   'form-0-email': other_user.email,
                                   'form-TOTAL_FORMS': 1,
                                   'form-MAX_NUM_FORMS': 1000,
                                   'form-MIN_NUM_FORMS': 0,
                                   'form-INITIAL_FORMS': 0,
                               })
        assert response.status_code == 200
```
but I get the following error message.

```
def get_success_url(self):
        """What page to show when the form was processed OK."""
        entity = self.object
        ns = self.request.resolver_match.namespace
>       return reverse(ns + ':entity_collaborators', args=[entity.url_type, entity.id])
E       AttributeError: 'Dataset' object has no attribute 'url_type'
```

**Comment by jonc125** [comment]
Ah, so I missed one point of difference between entity & dataset with the success URL, which means we will need this method to be defined different in each view. In the dataset version you don't need the `entity.url_type` argument, just the dataset ID is sufficient to define the URL. And you could call the local variable `dataset` rather than `entity` in that case for clarity!

**Comment by sroderick-g5sro** [comment]
> Interesting, because using the add_collaborator method directly we’ve tested and it works. It’d definitely be worth creating a failing test that uses the new view to add a collaborator so we have something more straightforward to debug. Possibly going via the form (which lives in entities) has caused Django to get confused about which app it’s looking for the edit_entity permission in?

The unit test that I have created seems to work as expected but I still get permissions error occuring when I run it in the browser.

**Comment by jonc125** [comment]
Annoyingly Travis has stopped running our tests due to their recent policy change :(  But running locally while there are 6 failing dataset tests, your new one is passing. When do you get the permissions error - when you view the form or when you submit it?

**Comment by sroderick-g5sro** [comment]
I get the error on submit

**Comment by jonc125** [comment]
The problem was missing DB migrations adding the permission in the datasets app. It works for tests because they create their own DB, and so don't need to run migrations. But testing live on the dev server does need the DB updated.

**Comment by sroderick-g5sro** [comment]
Great thanks

**Comment by jonc125** [comment]
@sroderick-g5sro could you resolve the merge conflicts, and I'll review after that?

**Comment by jonc125** [comment]
I can confirm the tests pass locally. (I've also requested more build minutes from Travis CI!)

**Comment by sroderick-g5sro** [comment]
Ok hopefully I have got the merge correct. I have run the tests locally and they all pass.

**Review by jonc125 — APPROVED** [review]
Works well here too!

**Pull request merged**